### PR TITLE
[Execution] Refactor Checker Engine

### PR DESCRIPTION
**Pull request by zhangchiqing**
Close https://github.com/onflow/flow-go/issues/5173

This PR refactors the checker engine to close https://github.com/onflow/flow-go/issues/5173

### CODE REVIEW

File: engine/execution/checker/core.go [hunk 1]
```diff
@@ -0,0 +1,170 @@
...
+
+	if lastExecutedHeight > lastFinal.Height {
+		// last executed block has not been finalized yet,
+		// can't check since unfinalized block is also unsealed, skip
+		return nil
+	}
+
+	// TODO: better to query seals from protocol state,
+	// switch to state.Final().LastSealed() when available
+	sealedExecuted, seal, err := c.findLatestSealedAtHeight(lastExecutedHeight)
```
Comment (peterargue) [inline engine/execution/checker/core.go:112]: what should happen if `err != nil`?

File: engine/execution/checker/core.go [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+	height, _, err := c.execState.GetHighestExecutedBlockID(context.Background())
+	if err != nil {
+		return 0, fmt.Errorf("could not get the last executed block: %w", err)
+	}
+	return height, nil
+}
+
+// findLatestSealedAtHeight finds the latest sealed block at the given height
+func (c *Core) findLatestSealedAtHeight(finalizedHeight uint64) (*flow.Header, *flow.Seal, error) {
+	_, seal, err := c.state.AtHeight(finalizedHeight).SealedResult()
```
Comment (peterargue) [inline engine/execution/checker/core.go:163]: same here

File: engine/execution/checker/core.go [hunk 3]
```diff
@@ -0,0 +1,170 @@
...
+	if err != nil {
+		return nil, nil, nil, fmt.Errorf("could not get the last sealed for the finalized block: %w", err)
+	}
+
+	lastSealed, err := c.state.AtBlockID(lastSeal.BlockID).Head()
+	if err != nil {
+		return nil, nil, nil, fmt.Errorf("could not get the last sealed block: %w", err)
+	}
+
+	return lastSealed, lastFinal, lastSeal, err
```
Comment (peterargue) [inline engine/execution/checker/core.go:149]: nit
```suggestion
	return lastSealed, lastFinal, lastSeal, nil
```

**Review by peterargue — APPROVED** [review]

File: engine/execution/checker/core.go [hunk 4]
```diff
@@ -0,0 +1,176 @@
...
+	e := &Core{
+		log:       logger.With().Str("engine", "checker").Logger(),
+		state:     state,
+		execState: execState,
+	}
+
+	return e
+}
+
+// The point of checker engine is to perform this following check
```
Comment (janezpodhostnik) [inline engine/execution/checker/core.go:41]:
```suggestion
// checkMyCommitWithSealedCommit is the main check of the checker engine
```

File: engine/execution/checker/core.go [hunk 5]
```diff
@@ -0,0 +1,176 @@
...
+	"github.com/rs/zerolog"
+
+	"github.com/onflow/flow-go/engine/execution/state"
+	"github.com/onflow/flow-go/model/flow"
+	"github.com/onflow/flow-go/state/protocol"
+	"github.com/onflow/flow-go/storage"
+)
+
+// trigger the check once every minute,
+const DefaultTimeInterval = time.Minute * 1
```
Comment (janezpodhostnik) [inline engine/execution/checker/core.go:18]:
```suggestion
// DefaultTimeInterval triggers the check once every minute,
const DefaultTimeInterval = time.Minute * 1
```
This belongs in the `engine.go` I think.

File: engine/execution/checker/core.go [hunk 6]
```diff
@@ -0,0 +1,176 @@
...
+
+// The checker engine checks if the execution result matches the sealed result.
+type Core struct {
```
Comment (janezpodhostnik) [inline engine/execution/checker/core.go:21]:
```suggestion
// Core is the core logic of the checker engine that checks if the execution result matches the sealed result.
type Core struct {
```

File: engine/execution/checker/core.go [hunk 7]
```diff
@@ -0,0 +1,176 @@
...
+		select {
+		case <-ticker.C:
+			err := c.runCheck()
+			if err != nil {
+				return err
+			}
+		case <-ctx.Done():
+			return nil
+		}
+	}
```
Comment (janezpodhostnik) [inline engine/execution/checker/core.go:80]: It might be cleaner to do this part in the engine.go

File: engine/execution/checker/core.go [hunk 8]
```diff
@@ -0,0 +1,176 @@
...
+}
+
+// it skips when the last sealed has not been executed, and last executed has not been finalized.
+func (c *Core) runCheck() error {
```
Comment (janezpodhostnik) [inline engine/execution/checker/core.go:84]:
```suggestion
// RunCheck skips when the last sealed has not been executed, and last executed has not been finalized.
func (c *Core) RunCheck() error {
```

File: engine/execution/checker/core.go [hunk 9]
```diff
@@ -0,0 +1,176 @@
...
+	// find last sealed block
+	lastSealedBlock, lastFinal, seal, err := c.findLastSealedBlock()
+	if err != nil {
+		return err
+	}
+
+	mycommitAtLastSealed, err := c.execState.StateCommitmentByBlockID(lastSealedBlock.ID())
```
Comment (janezpodhostnik) [inline engine/execution/checker/core.go:91]:
```suggestion
	myCommitAtLastSealed, err := c.execState.StateCommitmentByBlockID(lastSealedBlock.ID())
```

**Review by janezpodhostnik — APPROVED** [review]
Looks good. Just left comments about some cosmetic issues

File: engine/execution/checker/core.go (hunk 8 above)
Comment (zhangchiqing) [reply engine/execution/checker/core.go:84]: Any particular reason to make this method public?

File: engine/execution/checker/core.go (hunk 7 above)
Comment (zhangchiqing) [reply engine/execution/checker/core.go:80]: Good point

**Comment by zhangchiqing** [comment]
I added test cases for Core.go as well as a flag to enable/disable the checker (default is enabled) since the PR was approved.

**Review by janezpodhostnik — APPROVED** [review]

**Pull request merged**